source/connection_monitor.c

```diff
@@ -0,0 +1,222 @@
+        (current_outgoing_stream_id != 0 && current_outgoing_stream_id == impl->last_outgoing_stream_id);
+
+    impl->last_outgoing_stream_id = current_outgoing_stream_id;
+    impl->last_incoming_stream_id = current_incoming_stream_id;
+    impl->last_measured_throughput = bytes_per_second;
+
+    if (!check_throughput) {
+        AWS_LOGF_TRACE(AWS_LS_IO_CHANNEL, "id=%p: channel throughput does not need to be checked", (void *)channel);
+        impl->throughput_failure_time_ms = 0;
+        return;
+    }
+
+    if (bytes_per_second >= impl->options.minimum_throughput_bytes_per_second) {
+        impl->throughput_failure_time_ms = 0;
+        return;
+    }
+
+    impl->throughput_failure_time_ms =
+        aws_add_u64_saturating(impl->throughput_failure_time_ms, max_pending_io_interval_ms);
+
+    AWS_LOGF_INFO(
+        AWS_LS_IO_CHANNEL,
+        "id=%p: Channel low throughput warning.  Currently %" PRIu64 " milliseconds of consecutive failure time",
+        (void *)channel,
+        impl->throughput_failure_time_ms);
+
+    uint64_t maximum_failure_time_ms = aws_timestamp_convert(
+        impl->options.allowable_throughput_failure_interval_seconds, AWS_TIMESTAMP_SECS, AWS_TIMESTAMP_MILLIS, NULL);
+    if (impl->throughput_failure_time_ms <= maximum_failure_time_ms) {
+        return;
+    }
+
+    AWS_LOGF_INFO(
+        AWS_LS_IO_CHANNEL,
+        "id=%p: Channel low throughput threshold exceeded (< %" PRIu64
+        " bytes per second for more than %u seconds).  Shutting down.",
+        (void *)channel,
+        impl->options.minimum_throughput_bytes_per_second,
+        impl->options.allowable_throughput_failure_interval_seconds);
+
+    aws_channel_shutdown(channel, AWS_ERROR_HTTP_CHANNEL_THROUGHPUT_FAILURE);
+}
+
+static void s_destroy(struct aws_crt_statistics_handler *handler) {
+    if (handler == NULL) {
+        return;
+    }
+
+    aws_mem_release(handler->allocator, handler);
+}
+
+static uint64_t s_get_report_interval_ms(struct aws_crt_statistics_handler *handler) {
+    (void)handler;
+
+    return 1000;
+}
+
+static struct aws_crt_statistics_handler_vtable s_http_connection_monitor_vtable = {
+    .process_statistics = s_process_statistics,
+    .destroy = s_destroy,
+    .get_report_interval_ms = s_get_report_interval_ms,
+};
+
+struct aws_crt_statistics_handler *aws_crt_statistics_handler_new_http_connection_monitor(
+    struct aws_allocator *allocator,
+    struct aws_http_connection_monitoring_options *options) {
+    struct aws_crt_statistics_handler *handler = NULL;
+    struct aws_statistics_handler_http_connection_monitor_impl *impl = NULL;
+
+    if (!aws_mem_acquire_many(
+            allocator,
+            2,
+            &handler,
+            sizeof(struct aws_crt_statistics_handler),
+            &impl,
+            sizeof(struct aws_statistics_handler_http_connection_monitor_impl))) {
+        return NULL;
+    }
+
+    AWS_ZERO_STRUCT(*handler);
+    AWS_ZERO_STRUCT(*impl);
```

Comment:
style/opinion: you could just embed the `aws_crt_statistics_handler` within the impl struct. Then you can just use one calloc instead of acquire_many and 2 ZERO_STRUCTS

Reply:
Tbh I find that approach confusing.